From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Tue, 29 Nov 2022 12:55:45 +1000
Subject: Xtest: disallow GenericEvents in XTestSwapFakeInput
Origin: https://gitlab.freedesktop.org/xorg/xserver/commit/936d34bdff4c479ccd0405fc221ff8e4c6c7014d
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2022-46340
Bug-Debian: https://bugs.debian.org/1026071

XTestSwapFakeInput assumes all events in this request are
sizeof(xEvent) and iterates through these in 32-byte increments.
However, a GenericEvent may be of arbitrary length longer than 32 bytes,
so any GenericEvent in this list would result in subsequent events to be
misparsed.

Additional, the swapped event is written into a stack-allocated struct
xEvent (size 32 bytes). For any GenericEvent longer than 32 bytes,
swapping the event may thus smash the stack like an avocado on toast.

Catch this case early and return BadValue for any GenericEvent.
Which is what would happen in unswapped setups anyway since XTest
doesn't support GenericEvent.

CVE-2022-46340, ZDI-CAN 19265

This vulnerability was discovered by:
Jan-Niklas Sohn working with Trend Micro Zero Day Initiative

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
Acked-by: Olivier Fourdan <ofourdan@redhat.com>
(cherry picked from commit b320ca0ffe4c0c872eeb3a93d9bde21f765c7c63)
---
 Xext/xtest.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: xorg-server-1.20.4/Xext/xtest.c
===================================================================
--- xorg-server-1.20.4.orig/Xext/xtest.c	2022-12-30 01:48:39.606987526 +0100
+++ xorg-server-1.20.4/Xext/xtest.c	2022-12-30 01:48:39.606987526 +0100
@@ -501,10 +501,11 @@
 
     nev = ((req->length << 2) - sizeof(xReq)) / sizeof(xEvent);
     for (ev = (xEvent *) &req[1]; --nev >= 0; ev++) {
+        int evtype = ev->u.u.type & 0x177;
         /* Swap event */
-        proc = EventSwapVector[ev->u.u.type & 0177];
+        proc = EventSwapVector[evtype];
         /* no swapping proc; invalid event type? */
-        if (!proc || proc == NotImplemented) {
+        if (!proc || proc == NotImplemented || evtype == GenericEvent) {
             client->errorValue = ev->u.u.type;
             return BadValue;
         }
